contents/javascript/Javascript-promises.md

```diff
@@ -0,0 +1,164 @@
+
+To illustrate this problem in code, we'll use `readFile()` from the `fs` package in Node.js as an example. This is an asynchronous function takes in 3 parameters, a callback and a time (in milliseconds) to wait before the callback is executed. A callback is a function that is passed as a parameter to another function (call this A), and it will be executed after A finishes.
+
+In the following code snippet, you would probably expect `console.log(x)` to print `I have been updated`. Unfortunately, it prints `I have not been updated.` Give it a try in Google Chrome's developer console.
+
+```javascript
+
+var x = "I have not been updated";
+
+setTimeout(function (){
+    x = "I have been updated";
+}, 1000);
+
+console.log(x);
+
+```
+
+To fix this, we can simply shift the `console.log(x)` into the callback to get the expected result. It will now print the correct value.
+
+```javascript
+
+var x = "I have not been updated";
+
+setTimeout(function (){
+    x = "I have been updated";
+    console.log(x);
+}, 1000);
+
+```
+
+However, this fix will only go so far. If we had another `setTimeout()` that depended on the result of the earlier `setTimeout()`, we would have to nest the functions within each other which would make for unreadable code as follows. 
+
+```javascript
+
+var x = "I have not been updated";
+
+setTimeout(function (){
+    x = "I have been updated";
+    setTimeout(function () {
+        x = "I have been updated 2 times.";
+        setTimeout(function () {
+            x = "I have been updated 3 times.";
+            // and so on...
+        }, 1000);
+    }, 1000);
+}, 1000);
+
+```
+
+That's pretty ugly and it would be difficult for anyone to read your code and to understand what exactly is going on.
+
+We can instead use Promises for cleaner code that would be easier to write and debug.
+
+## What is a Promise?
+
+The following example is adapted from [JavaScript Promises for Dummies](https://scotch.io/tutorials/javascript-promises-for-dummies).
+
+Promises in Javascript behave the same way as Promises do in real life. Imagine that your friend promises to return you money that you have lent him. At the time this promise was made, you would not know if your friend would really return your money. Your friend could either return your money on time, or he could just not do it.
+These situations correspond to the 3 states of Promises in Javascript.
+
+| State | Description |
+| ------ | ----------- |
+|Pending | You don't know if he would return your money. |
+|Fulfilled | He returned your money. |
+|Rejected | He refused to return your money. |
+
+
+Promises provide the power to control the execution flow of your program by allowing you to wait for a asynchronous operation to complete before carrying on with the rest of the program execution.
```

Comment:
This statement sounds like "promises provide a way to do things synchronously" (because it says that the rest of the program will not execute until the async code is executed). I don't know promises myself, but the statement didn't sound right to me. Did you mean "promises allows you to specify how the execution of _some part_ of the code to depend on the status of an asynchronous code" or something of the sort.